From acbd5a2db28a27c1af85e4e3bcb7fe9b17639c5a Mon Sep 17 00:00:00 2001
From: Padmarao Begari <padmarao.begari@microchip.com>
Date: Tue, 18 Jan 2022 12:04:43 +0530
Subject: [PATCH 2/9] U-Boot v2022.01 with high memory 0x1000000000

Enable the DDR 1GB high memory(0x1000000000), the MMC 64-bit DMA,
Host Controller version 4 and update ADMA descriptor table for
64-bit address.

Signed-off-by: Padmarao Begari <padmarao.begari@microchip.com>
Fixed DDR size &
Signed-off-by: Conor Dooley <conor.dooley@microchip.com>
---
 arch/riscv/cpu/generic/dram.c                |  3 ++-
 arch/riscv/dts/microchip-mpfs-icicle-kit.dts | 12 ++----------
 board/microchip/mpfs_icicle/Kconfig          |  4 ++--
 configs/microchip_mpfs_icicle_defconfig      |  2 +-
 drivers/mmc/sdhci-adma.c                     |  1 +
 drivers/mmc/sdhci.c                          |  7 ++++++-
 include/configs/microchip_mpfs_icicle.h      | 14 +++++++-------
 include/sdhci.h                              |  2 ++
 8 files changed, 23 insertions(+), 22 deletions(-)

diff --git a/arch/riscv/cpu/generic/dram.c b/arch/riscv/cpu/generic/dram.c
index 1fdc7837b8..584e656169 100644
--- a/arch/riscv/cpu/generic/dram.c
+++ b/arch/riscv/cpu/generic/dram.c
@@ -23,6 +23,7 @@ int dram_init_banksize(void)
 
 ulong board_get_usable_ram_top(ulong total_size)
 {
+#if !defined(CONFIG_TARGET_MICROCHIP_ICICLE)
 	/*
 	 * Ensure that we run from first 4GB so that all
 	 * addresses used by U-Boot are 32bit addresses.
@@ -33,6 +34,6 @@ ulong board_get_usable_ram_top(ulong total_size)
 	 */
 	if (gd->ram_top >= SZ_4G)
 		return SZ_4G - 1;
-
+#endif
 	return gd->ram_top;
 }
diff --git a/arch/riscv/dts/microchip-mpfs-icicle-kit.dts b/arch/riscv/dts/microchip-mpfs-icicle-kit.dts
index 287ef3d23b..1da1d5139a 100644
--- a/arch/riscv/dts/microchip-mpfs-icicle-kit.dts
+++ b/arch/riscv/dts/microchip-mpfs-icicle-kit.dts
@@ -79,16 +79,9 @@
 		sync-mode = <3>;
 	};
 
-	ddrc_cache_lo: memory@80000000 {
+	ddrc_cache: memory@1000000000 {
 		device_type = "memory";
-		reg = <0x0 0x80000000 0x0 0x2e000000>;
-		clocks = <&clkcfg CLK_DDRC>;
-		status = "okay";
-	};
-
-	ddrc_cache_hi: memory@1000000000 {
-		device_type = "memory";
-		reg = <0x10 0x0 0x0 0x40000000>;
+		reg = <0x10 0x0 0x0 0x76000000>;
 		clocks = <&clkcfg CLK_DDRC>;
 		status = "okay";
 	};
@@ -100,7 +93,6 @@
 
 &mmc {
 	status = "okay";
-
 	bus-width = <4>;
 	disable-wp;
 	cap-mmc-highspeed;
diff --git a/board/microchip/mpfs_icicle/Kconfig b/board/microchip/mpfs_icicle/Kconfig
index 092e411215..4f2a236fbb 100644
--- a/board/microchip/mpfs_icicle/Kconfig
+++ b/board/microchip/mpfs_icicle/Kconfig
@@ -13,8 +13,8 @@ config SYS_CONFIG_NAME
 	default "microchip_mpfs_icicle"
 
 config SYS_TEXT_BASE
-	default 0x80000000 if !RISCV_SMODE
-	default 0x80200000 if RISCV_SMODE
+	default 0x1000000000 if !RISCV_SMODE
+	default 0x1000200000 if RISCV_SMODE
 
 config BOARD_SPECIFIC_OPTIONS # dummy
 	def_bool y
diff --git a/configs/microchip_mpfs_icicle_defconfig b/configs/microchip_mpfs_icicle_defconfig
index e51afc17e2..fd0c3dd47c 100644
--- a/configs/microchip_mpfs_icicle_defconfig
+++ b/configs/microchip_mpfs_icicle_defconfig
@@ -7,7 +7,7 @@ CONFIG_TARGET_MICROCHIP_ICICLE=y
 CONFIG_ARCH_RV64I=y
 CONFIG_RISCV_SMODE=y
 CONFIG_DISTRO_DEFAULTS=y
-CONFIG_SYS_LOAD_ADDR=0x80200000
+CONFIG_SYS_LOAD_ADDR=0x1000200000
 CONFIG_FIT=y
 CONFIG_DISPLAY_CPUINFO=y
 CONFIG_DISPLAY_BOARDINFO=y
diff --git a/drivers/mmc/sdhci-adma.c b/drivers/mmc/sdhci-adma.c
index 2ec057fbb1..1406643bd8 100644
--- a/drivers/mmc/sdhci-adma.c
+++ b/drivers/mmc/sdhci-adma.c
@@ -24,6 +24,7 @@ static void sdhci_adma_desc(struct sdhci_adma_desc *desc,
 	desc->addr_lo = lower_32_bits(addr);
 #ifdef CONFIG_DMA_ADDR_T_64BIT
 	desc->addr_hi = upper_32_bits(addr);
+	desc->rsvd1 = 0;
 #endif
 }
 
diff --git a/drivers/mmc/sdhci.c b/drivers/mmc/sdhci.c
index 766e4a6b0c..b7552c8755 100644
--- a/drivers/mmc/sdhci.c
+++ b/drivers/mmc/sdhci.c
@@ -77,6 +77,7 @@ static void sdhci_prepare_dma(struct sdhci_host *host, struct mmc_data *data,
 	dma_addr_t dma_addr;
 	unsigned char ctrl;
 	void *buf;
+	u16 hv64 = 0;
 
 	if (data->flags == MMC_DATA_READ)
 		buf = data->dest;
@@ -85,8 +86,12 @@ static void sdhci_prepare_dma(struct sdhci_host *host, struct mmc_data *data,
 
 	ctrl = sdhci_readb(host, SDHCI_HOST_CONTROL);
 	ctrl &= ~SDHCI_CTRL_DMA_MASK;
-	if (host->flags & USE_ADMA64)
+	if (host->flags & USE_ADMA64) {
 		ctrl |= SDHCI_CTRL_ADMA64;
+		hv64 = sdhci_readw(host, SDHCI_HOST_CONTROL2);
+		hv64 |= SDHCI_CTRL_HV4E_A64B_ENABLE;
+		sdhci_writew(host, hv64, SDHCI_HOST_CONTROL2);
+	}
 	else if (host->flags & USE_ADMA)
 		ctrl |= SDHCI_CTRL_ADMA32;
 	sdhci_writeb(host, ctrl, SDHCI_HOST_CONTROL);
diff --git a/include/configs/microchip_mpfs_icicle.h b/include/configs/microchip_mpfs_icicle.h
index 655c8d6af5..0b7e56af93 100644
--- a/include/configs/microchip_mpfs_icicle.h
+++ b/include/configs/microchip_mpfs_icicle.h
@@ -9,12 +9,12 @@
 
 #include <linux/sizes.h>
 
-#define CONFIG_SYS_SDRAM_BASE       0x80000000
+#define CONFIG_SYS_SDRAM_BASE       0x1000000000
 #define CONFIG_SYS_INIT_SP_ADDR     (CONFIG_SYS_SDRAM_BASE + SZ_2M)
 
 #define CONFIG_SYS_BOOTM_LEN        SZ_64M
 
-#define CONFIG_STANDALONE_LOAD_ADDR 0x80200000
+#define CONFIG_STANDALONE_LOAD_ADDR 0x1000200000
 
 /* Environment options */
 
@@ -26,11 +26,11 @@
 
 #define CONFIG_EXTRA_ENV_SETTINGS \
 	"bootm_size=0x10000000\0" \
-	"kernel_addr_r=0x84000000\0" \
-	"fdt_addr_r=0x88000000\0" \
-	"scriptaddr=0x88100000\0" \
-	"pxefile_addr_r=0x88200000\0" \
-	"ramdisk_addr_r=0x88300000\0" \
+	"kernel_addr_r=0x1004000000\0" \
+	"fdt_addr_r=0x1008000000\0" \
+	"scriptaddr=0x1008100000\0" \
+	"pxefile_addr_r=0x1008200000\0" \
+	"ramdisk_addr_r=0x1008300000\0" \
 	BOOTENV
 
 #endif /* __CONFIG_H */
diff --git a/include/sdhci.h b/include/sdhci.h
index c718dd7206..8312ce3b75 100644
--- a/include/sdhci.h
+++ b/include/sdhci.h
@@ -165,6 +165,7 @@
 #define  SDHCI_CTRL_EXEC_TUNING	0x0040
 #define  SDHCI_CTRL_TUNED_CLK	0x0080
 #define  SDHCI_CTRL_PRESET_VAL_ENABLE	0x8000
+#define  SDHCI_CTRL_HV4E_A64B_ENABLE	0x3000
 
 #define SDHCI_CAPABILITIES	0x40
 #define  SDHCI_TIMEOUT_CLK_MASK	0x0000003F
@@ -302,6 +303,7 @@ struct sdhci_adma_desc {
 	u32 addr_lo;
 #ifdef CONFIG_DMA_ADDR_T_64BIT
 	u32 addr_hi;
+	u32 rsvd1;
 #endif
 } __packed;
 
-- 
2.25.1

